# CHOCOLATE DATA ANALYSIS


***
***
<div class="alert alert-block alert-warning">

<p style="font-family:Rockwell">
        <b>This is an analysis project based on chocolate dataset found on kaggle.com.
            <br> The data consists of the following columns :</b>
</p>
<p style="font-family:Optima">
        <br><b>company : </b> Provides company name of the company producing the chocolate bar.
        <br><b>company_location : </b> Provides the name of country where the chocolate bar producing company is located.
        <br><b>country_of_bean_origin : </b> Provides the name of country where the chocolate <b>bean used in chocolate bar</b> is produced.
        <br><b>specific_bean_origin_or_bar_name : </b> Name of the chocolate bar or bean used for production.
        <br><b>cocoa_percent : </b> Percentage of Cocoa used in the chocolate bar.
        <br><b>rating : </b> Rating in stars (*) collected for the chocolate bar.
        <br><b>rating : </b> Rating in stars (*) collected for the chocolate bar.
        <br><b>cocoa_butter : </b> Provides information if cocoa_butter is one of the ingredients of the chocolate bar.
        <br><b>vanilla : </b> Provides information if vanilla is one of the ingredients of the chocolate bar.
        <br><b>lecinthin : </b> Provides information if lecinthin is one of the ingredients of the chocolate bar.
        <br><b>salt : </b> Provides information if salt is one of the ingredients of the chocolate bar.
        <br><b>sugar : </b> Provides information if sugar is one of the ingredients of the chocolate bar.
        <br><b>sweetener_without_sugar : </b> Provides information if sweetener_without_sugar was used for production of the chocolate bar
        <br><b>first_taste : </b> Provides information about the primary taste of the chocolate bar.
        <br><b>second_taste : </b> Provides information about the secondary taste of the chocolate bar.
        <br><b>third_taste : </b> Provides information about the tertiary taste of the chocolate bar.
        <br><b>fourth_taste : </b> Provides information about the 4 taste of the chocolate bar.
</p>

</div>

 ***


***
<div class="alert alert-block alert-info">
<br> We will be exploring the following topics :
<br>
    
<strong>1. [Chocolate Bars and Rating Distribution.](#Chocolate-Bars-and-Rating-Distribution)</strong>
<br><br>
<strong>2. [Most liked and diskiled taste first taste, second taste.](#Most-liked-taste-and-not-at-all-liked-tastes)</strong>
<br><br>
<strong>3. [Explore Ingredients Distribution.](#Explore-Ingredients-distribution)</strong>
    <li> [Cocoa Distribution.](#Cocoa-Distribution)</li>
         <li> [Other Ingredients Analysis.](#Other-Ingredients-Analysis)</li>
<br>
<strong>4. [Chocolate BEAN production analysis.](#Countries-producing-chocolate-BEAN,-their-ratings-and-the-number-of-beans-that-they-produce)</strong>
<br><br>
<strong>5. [Chocolate BAR Production Analysis](#Chocolate-BAR-Production-Analysis)</strong>

</div>

***


In [647]:
# print all the outputs in a cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
%config IPCompleter.greedy=True

# import the basic libraries.
import pandas as pd
import numpy as np


In [648]:
# Set display setting to show 101 columns and 1000 rows
pd.set_option("display.max_rows", 1000)
pd.set_option("display.max_columns", 101)

In [2]:
#Create dataframes from the csv files
chocolate=pd.read_csv("chocolate.csv")

In [102]:
chocolate.head()

Unnamed: 0   ref   company company_location  review_date  \
0           0  2454      5150            U.S.A         2019   
1           1  2458      5150            U.S.A         2019   
2           2  2454      5150            U.S.A         2019   
3           3   797  A. Morin           France         2012   
4           4   797  A. Morin           France         2012   

  country_of_bean_origin specific_bean_origin_or_bar_name  cocoa_percent  \
0             Madagascar           Bejofo Estate, batch 1           76.0   
1     Dominican republic                  Zorzal, batch 1           76.0   
2               Tanzania            Kokoa Kamili, batch 1           76.0   
3                   Peru                             Peru           63.0   
4                Bolivia                          Bolivia           70.0   

   rating  counts_of_ingredients      beans       cocoa_butter  \
0    3.75                      3  have_bean  have_cocoa_butter   
1    3.50                      3  have_bean  have_cocoa_butter   
2    3.25                      3  have_bean  have_cocoa_butter   
3    3.75                      4  have_bean  have_cocoa_butter   
4    3.50                      4  have_bean  have_cocoa_butter   

           vanilla           lecithin           salt       sugar  \
0  have_not_vanila  have_not_lecithin  have_not_salt  have_sugar   
1  have_not_vanila  have_not_lecithin  have_not_salt  have_sugar   
2  have_not_vanila  have_not_lecithin  have_not_salt  have_sugar   
3  have_not_vanila      have_lecithin  have_not_salt  have_sugar   
4  have_not_vanila      have_lecithin  have_not_salt  have_sugar   

            sweetener_without_sugar first_taste second_taste third_taste  \
0  have_not_sweetener_without_sugar       cocoa   blackberry   full body   
1  have_not_sweetener_without_sugar       cocoa      vegetal      savory   
2  have_not_sweetener_without_sugar  rich cocoa        fatty      bready   
3  have_not_sweetener_without_sugar      fruity        melon      roasty   
4  have_not_sweetener_without_sugar     vegetal        nutty         NaN   

  fourth_taste company_location_new  
0          NaN                U.S.A  
1          NaN                U.S.A  
2          NaN                U.S.A  
3          NaN               France  
4          NaN               France

*Clean data: Fix country name typos and capitalize the names*


In [103]:
chocolate['company_location']=chocolate['company_location'].str.title()

In [104]:
chocolate['country_of_bean_origin']=chocolate['country_of_bean_origin'].str.title()

In [593]:
def clean_location_name(row):
    if row['country_of_bean_origin']=='Bolvia':
        return "Bolivia"
    elif row['country_of_bean_origin']=='Burma':
        return "Myanmar"
    elif row['country_of_bean_origin']=='Dr Congo':
        return "Democratic Republic of Congo"
    elif row['country_of_bean_origin']=='Sao Tome & Principe':
        return "Sao Tome and Principe"
    elif row['country_of_bean_origin']=='Sao Tome':
        return "Sao Tome and Principe"
    elif row['country_of_bean_origin']=='Principe':
        return "Sao Tome and Principe"
    elif row['country_of_bean_origin']=='St. Lucia':
        return "Saint Lucia"
    elif row['country_of_bean_origin']=='St.Vincent-Grenadines':
        return "Saint Vincent and the Grenadines"
    elif row['country_of_bean_origin']=='Sulawesi':
        return "Indonesia"
    elif row['country_of_bean_origin']=='Sumatra':
        return "Indonesia"
    elif row['country_of_bean_origin']=='Tobago':
        return "Trinidad and Tobago"
    elif row['country_of_bean_origin']=='Trinidad':
        return "Trinidad and Tobago"
    elif row['country_of_bean_origin']=='Scotland':
        return "U.K."
    elif row['country_of_bean_origin']=='Ireland':
        return "U.K."
    else:
        return row['country_of_bean_origin']

In [610]:
def clean_location_name_2(row):
    if row['company_location']=='Bolvia':
        return "Bolivia"
    elif row['company_location']=='Burma':
        return "Myanmar"
    elif row['company_location']=='Dr Congo':
        return "Democratic Republic of Congo"
    elif row['company_location']=='Sao Tome & Principe':
        return "Sao Tome and Principe"
    elif row['company_location']=='Sao Tome':
        return "Sao Tome and Principe"
    elif row['company_location']=='Principe':
        return "Sao Tome and Principe"
    elif row['company_location']=='St. Lucia':
        return "Saint Lucia"
    elif row['company_location']=='St.Vincent-Grenadines':
        return "Saint Vincent and the Grenadines"
    elif row['company_location']=='Sulawesi':
        return "Indonesia"
    elif row['company_location']=='Sumatra':
        return "Indonesia"
    elif row['company_location']=='Tobago':
        return "Trinidad and Tobago"
    elif row['company_location']=='Trinidad':
        return "Trinidad and Tobago"
    elif row['company_location']=='Scotland':
        return "U.K."
    elif row['company_location']=='Ireland':
        return "U.K."
    else:
        return row['company_location']

Create new columns with cleaned country names

In [611]:
chocolate['country_of_bean_origin_new']=chocolate.apply(clean_location_name,axis=1)

In [612]:
chocolate['company_location_new']=chocolate.apply(clean_location_name_2,axis=1)

### Chocolate Bars and Rating Distribution


In [649]:
# Group by bar name and average the ratings, to get the avg rating for each bar


In [291]:
popular_bar=chocolate.groupby('specific_bean_origin_or_bar_name')[['rating']].agg({'mean','count'})
popular_bar.columns=popular_bar.columns.droplevel(level=0)
popular_bar=popular_bar.rename(columns={'mean':'Avg Rating','count':'# of bars'})
popular_bar=popular_bar.sort_values(by='Avg Rating',ascending=False)
popular_bar



Avg Rating  # of bars
specific_bean_origin_or_bar_name                                       
Solomon Is.,batch 17158, N.Vela                     4.000000          1
Kulili P., 2013                                     4.000000          1
Choroni, Aragua, Finca Torres Farm                  4.000000          1
San Juan de Cheni                                   4.000000          1
Bachelor's Hall E., St. Thomas Parish, batch bh1    4.000000          1
Chuao, #217, DR, MC, batch 11-021                   4.000000          1
Maracaibo, El Vigia                                 4.000000          1
Maracaibo, El Rosario                               4.000000          1
Manjari                                             4.000000          1
Asante                                              4.000000          1
Madre de Dios                                       4.000000          1
Claudio Corallo w/ nibs                             4.000000          1
Silvestre, Batch 7, 2013                            4.000000          1
Madagascar, 100% criollo                            4.000000          1
Piura, Illanka, Quemazon                            4.000000          1
Somia, 2013                                         4.000000          1
Sur del Lago, Amiari Meridena, Zulia, w/ nibs       4.000000          1
Tangara                                             4.000000          1
Coopertiva Amazona w/ nibs                          4.000000          1
Three Amigos(Chuao, Wild Bolivia, D.R.)             4.000000          1
Tien Giang, Black S., batch VIT60420.0              4.000000          1
Corona Arriba                                       4.000000          1
Coto Brus, Heirloom, Batch 1                        4.000000          1
Marcial, single Cote, 2012                          4.000000          1
Markham Valley, #220, MR, MC                        4.000000          1
Maya Mtn, Batch 454, Heirloom                       4.000000          1
Bellavista Coop, #225, LR, MC, CG Exclusive         4.000000          1
Phantom                                             4.000000          1
Peruvian                                            4.000000          1
Porcelana, Pedegral                                 4.000000          1
Porcelana, Sorotaima,Machiques,batch pcl001         4.000000          1
Porcelana, Tabasco, La Joya                         4.000000          1
Porcelana, Tabasco, Limited Ed.                     4.000000          1
Porcelana, Zulia                                    4.000000          1
Cabosse, Venezuela, Java                            4.000000          1
Pablino                                             4.000000          1
Otucan, Grand Cru                                   4.000000          1
Sambirano, Ambanja, Madagascar                      4.000000          1
Bejofo, Sambirano V., Akesson E.                    4.000000          1
Rio Dulce, Xoco                                     4.000000          1
Rio Peripa H.                                       4.000000          1
Ocumare 67, Puertofino                              4.000000          1
Bali, Sukrama Bros. Farm, Melaya, 62hr C            4.000000          1
Sambirano Valley, #216, MR, LC                      4.000000          1
Carenero Superior, Urrutia, Barlovento              4.000000          1
Morobe                                              4.000000          1
Sambirano, 2011 harvest, batch 1                    4.000000          1
Coto Brus, Terciopelo                               4.000000          1
Piura, Choc. Garage Exclusive                       4.000000          1
Xoconusco, Chiapas, cacao Real                      4.000000          1
ABOCFA Coop                                         4.000000          1
Alto Beni, Cru Savage                               4.000000          1
Equator, Peru, Dom. Rep                             4.000000          1
Indio Rojo                                          4.000000          1
Wild Bolivia, Sucr

In [293]:
popular_bar=popular_bar.reset_index()
popular_bar[['Avg Rating']].describe()

Avg Rating
count  1398.000000
mean      3.224331
std       0.416174
min       1.500000
25%       3.000000
50%       3.250000
75%       3.500000
max       4.000000

In [299]:
# Group by 'Avg Rating column to be able to determine the number of chocolate bars belonging to particular avg rating groups'
num_bars_ratings=popular_bar.groupby('Avg Rating')[['# of bars']].sum().reset_index()

In [304]:
len(num_bars_ratings)

72

In [359]:
#DEfine a function to group the ratings 
def group_ratings(row):
    if (row['Avg Rating']>=0 and row['Avg Rating']<=1.5):
        val="0 to 1.5 stars"
    elif (row['Avg Rating']>1.5 and row['Avg Rating']<2.0):
        val="1.5 to 2 stars"
    elif (row['Avg Rating']>=2.0 and row['Avg Rating']<2.5):
        val="2 to 2.5 stars"
    elif (row['Avg Rating']>=2.5 and row['Avg Rating']<3.0):
        val="2.5 to 3 stars"
    elif (row['Avg Rating']>=3.0 and row['Avg Rating']<3.5):
        val="3 to 3.5 stars"
    elif (row['Avg Rating']>=3.5 and row['Avg Rating']<=4):
        val="3.5 to 4 stars"
    return val

In [360]:
#apply function
num_bars_ratings['Ratings Group']=num_bars_ratings.apply(group_ratings,axis=1)

In [362]:
num_bars_ratings2=num_bars_ratings.groupby('Ratings Group')[['# of bars']].sum().reset_index()

In [370]:

num_bars_ratings2

Ratings Group  # of bars
0  0 to 1.5 stars          3
1  1.5 to 2 stars          4
2  2 to 2.5 stars         26
3  2.5 to 3 stars        343
4  3 to 3.5 stars       1254
5  3.5 to 4 stars        594

In [684]:
import plotly.graph_objects as go

# ---- Plot box plot to understand the rating distribution ---
fig = go.Figure()
# Use x instead of y argument for horizontal plot
fig=fig.add_trace(go.Box(x=popular_bar['Avg Rating'],name='Avg Rating (in stars)',marker_color = 'lightseagreen',
                        boxmean=True))
fig=fig.update_layout(title_text="Box plot showing distribution of Chocolate bar Ratings",
    height=300,
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="darkolivegreen",)

fig.show()

#--- Plot pie chart showing number of chocolate bars in each Avg. Rating groups --
sunflowers_colors = ['rgb(175, 49, 35)', 'rgb(205, 152, 36)',
                      'rgb(124, 103, 37)','rgb(146, 123, 21)','rgb(99, 79, 37)','rgb(129, 180, 179)']
labels = num_bars_ratings2['Ratings Group'].values
values = num_bars_ratings2['# of bars'].values

fig = go.Figure(data=[go.Pie(labels=labels, values=values, marker_colors=sunflowers_colors)])
fig=fig.update_layout(title_text="Avg Rating and number of bars ", height=450,
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="darkolivegreen",)

fig.show()



<div class="alert alert-block alert-info">

Looks like there are many chocolate bar that have a rating of 4 stars. 
26% of the bars, i.e 594 chocolate bars have a rating between 3.5 to 4 stars and the overall average rating appears to be 3.2 stars. 

</div>

### Most liked taste and not at all liked tastes

 _2.1 Best and worst first tastes_

In [650]:
# Group by First_taste to find the Avg rating and number of chocolate bars belonging to each first_taste
first_taste=chocolate.groupby('first_taste')[['rating']].agg({'mean','count'})
first_taste.columns=first_taste.columns.droplevel(level=0)
first_taste=first_taste.rename(columns={'mean':'Avg Rating','count':'# of bars'})
first_taste=first_taste.sort_values(by='Avg Rating',ascending=False)
first_taste

Avg Rating  # of bars
first_taste                                          
distinct lemon                    4.000000          1
ligt color                        4.000000          1
simple red berry                  4.000000          1
smoth                             4.000000          1
cocoa and coconut                 4.000000          1
cinamon and nutmeg                4.000000          1
chocolate covered banana          4.000000          1
cardamon                          4.000000          1
well defined                      4.000000          1
pronounced berry                  4.000000          1
sweet spice                       4.000000          1
blackberry and cinnamon           4.000000          1
large nibs                        4.000000          1
mild tart                         4.000000          2
peanut                            3.875000          2
med roast                         3.750000          1
strong cocoa                      3.750000          1
cocoa and fresh grapes            3.750000          1
mild cacao and fruit notes        3.750000          1
bright fruit                      3.750000          4
bright red fruit                  3.750000          1
grapefruit                        3.750000          1
intense grapes                    3.750000          1
intense fruity                    3.750000          1
apple                             3.750000          1
subtle red berry                  3.750000          1
subtle but refreshing             3.750000          1
clingy                            3.750000          1
cigar                             3.750000          1
red wine                          3.750000          1
spicy and fragrant                3.750000          1
strong spice                      3.750000          2
oranges                           3.750000          1
strawberries                      3.750000          1
distinct                          3.750000          1
strong nutty                      3.750000          1
strong malt                       3.750000          1
hay                               3.750000          1
bold cherry                       3.750000          1
intense orange                    3.750000          1
thick                             3.750000          1
mellow fruit                      3.750000          1
marshmallow                       3.750000          3
dense                             3.750000          1
dominant spice                    3.750000          1
mellow                            3.750000          2
deep                              3.750000          1
yellow fruit                      3.750000          1
dark forest berry                 3.750000          1
walnuts                           3.750000          1
mildy nutty                       3.750000          1
tropical fruit and mild smoke     3.750000          1
cocoa dominates                   3.750000          1
short                             3.750000          1
evolves black pepper to citrus    3.750000          1
cranberries                       3.750000          1
roasty cocoa                      3.750000          1
full bodied                       3.750000          1
black tea                         3.750000          1
blackberries                      3.750000          1
intense prune and berry           3.750000          1
well balanced                     3.750000          1
appealing smoke                   3.750000          1
delicate                          3.714286          7
balanced                          3.678571          7
robust                            3.666667          3
fig                               3.666667          3
dark berry                        3.656250          8
roasted nuts                      3.625000          4
cherry                            3.607143          7
simple                            3.600000          5
grapes                            3.583333          3
long lasting                      3.

_Top tastes :_

In [394]:
first_taste[first_taste['Avg Rating']==4.0]

Avg Rating  # of bars
first_taste                                    
distinct lemon                   4.0          1
ligt color                       4.0          1
simple red berry                 4.0          1
smoth                            4.0          1
cocoa and coconut                4.0          1
cinamon and nutmeg               4.0          1
chocolate covered banana         4.0          1
cardamon                         4.0          1
well defined                     4.0          1
pronounced berry                 4.0          1
sweet spice                      4.0          1
blackberry and cinnamon          4.0          1
large nibs                       4.0          1
mild tart                        4.0          2

_Tastes to avoid (or worst tastes):_ 

In [395]:
first_taste.tail(10)

Avg Rating  # of bars
first_taste                                
burnt up front              2.25          1
raw                         2.25          1
non descript                2.00          1
multiple off flavors        2.00          1
bitter coffee flavor        2.00          1
fresh nibs                  2.00          1
sandpaper                   2.00          1
malitol                     2.00          1
klingy                      1.50          1
chemical                    1.50          1

_2.2. Best and Worst second tastes :_


In [651]:
second_taste=chocolate.groupby('second_taste')[['rating']].agg({'mean','count'})
second_taste.columns=second_taste.columns.droplevel(level=0)
second_taste=second_taste.rename(columns={'mean':'Avg Rating','count':'# of bars'})
second_taste=second_taste.sort_values(by='Avg Rating',ascending=False)
#second_taste

_Top second tastes :_

In [397]:
second_taste[second_taste['Avg Rating']==4.0]

Avg Rating  # of bars
second_taste                          
yogurt                  4.0          1
unripened fruit         4.0          1
long and rich           4.0          1
intense pepper          4.0          1
homey                   4.0          1
raspberries             4.0          1
roasted cocoa           4.0          1
dense                   4.0          1
delicate cocoa          4.0          1
rounded                 4.0          1
burnt brownie           4.0          1
tart citrus             4.0          1
distinquished           4.0          1

_Some second tastes that you should never and never ever try if the first tastes belongs to the one the above flagged bad first tastes :_

In [398]:
second_taste.tail(10)

Avg Rating  # of bars
second_taste                             
some spice                2.00          1
some fruit                2.00          1
vanilla domintates        2.00          1
unrefined                 2.00          1
flat flavor               2.00          1
poor aftertaste           2.00          1
dominate off note         1.75          1
strong chemical           1.50          1
hint of fruit             1.50          1
strong off flavor         1.00          1

2.3 : Best and worst first and second tastes : 

In [652]:
first_second_taste=chocolate.groupby(['first_taste','second_taste'])[['rating']].agg({'mean','count'})
first_second_taste.columns=first_second_taste.columns.droplevel(level=0)
first_second_taste=first_second_taste.rename(columns={'mean':'Avg Rating','count':'# of bars'})
first_second_taste=first_second_taste.sort_values(by='Avg Rating',ascending=False)
#first_second_taste

In [400]:
first_second_taste[first_second_taste['Avg Rating']==4.0]

Avg Rating  # of bars
first_taste      second_taste                          
dark berry       honey                   4.0          1
floral           berry                   4.0          1
complex          spice                   4.0          1
                 raspberry               4.0          1
creamy           rounded                 4.0          1
intense          raisins                 4.0          1
creamy           sticky                  4.0          3
smoth            dense                   4.0          1
peanut butter    raspberry               4.0          1
dark berry       tropical                4.0          1
banana           pear                    4.0          1
complex          balanced                4.0          1
strong spice     intense pepper          4.0          1
peanut           coffee                  4.0          1
smooth           nutty                   4.0          1
delicate         hazelnut                4.0          1
                 nutty                   4.0          1
green            unripened fruit         4.0          1
caramel          nuts                    4.0          1
tart             roasted cocoa           4.0          1
                 red berry               4.0          1
creamy           raisin                  4.0          1
                 dried fruit             4.0          1
                 homey                   4.0          1
sweet spice      floral                  4.0          1
tangy            floral                  4.0          1
creamy           bright fruit            4.0          1
bright fruit     molasses                4.0          1
nuts             butterscotch            4.0          1
pronounced berry mild smoke              4.0          1
mild tart        wine                    4.0          1
                 long and rich           4.0          1
creamy           mint                    4.0          1
                 acidic                  4.0          1
                 nuts                    4.0          1
                 pistachio               4.0          1
tart             lemon                   4.0          1
banana           yogurt                  4.0          1
honey            deep cocoa              4.0          1
rich cocoa       tart citrus             4.0          1
simple           delicate cocoa          4.0          1
long lasting     cocoa                   4.0          1
well defined     creamy                  4.0          1
smokey           raspberries             4.0          1
oily             nut                     4.0          1
strawberry       cocoa                   4.0          1
simple red berry rich                    4.0          1
roasted nuts     burnt brownie           4.0          1
fatty            spicy                   4.0          1
nutty            tart                    4.0          1
balanced         cherry                  4.0          1
mild             licorice                4.0          1
marshmallow      distinquished           4.0          1
large nibs       spicy                   4.0          1
light color      complex                 4.0          1
balanced         nuts                    4.0          1
ligt color       creamy                  4.0          1
oily             subtle                  4.0          1
banana           dairy                   4.0          1

In [401]:
first_second_taste.tail(10)

Avg Rating  # of bars
first_taste second_taste                             
dry         flat flavor               2.00          1
fresh nibs  chocolate base off        2.00          1
dry         bitter                    2.00          1
sweet       vanilla domintates        2.00          1
perfume     chemical                  2.00          1
cocoa       dominate off note         1.75          1
klingy      hint of fruit             1.50          1
perfume     strong chemical           1.50          1
chemical    salt                      1.50          1
pastey      strong off flavor         1.00          1

Combine first_taste top flavors and second_taste top flavors (rating = 4.0) for graph purpose

In [664]:
first_taste=first_taste.reset_index()
first_taste['First/Second']="First Taste"

In [665]:
first_taste.rename(columns={'first_taste': 'Taste'},inplace=True)

In [666]:
second_taste=second_taste.reset_index()
second_taste['First/Second']="Second Taste"

In [667]:
second_taste.rename(columns={'second_taste': 'Taste'},inplace=True)

In [668]:
len(second_taste)+len(first_taste)

935

In [669]:
#Combine by appending above 2 dataframes
combined_taste=first_taste.append(second_taste)

In [670]:
len(combined_taste)

935

In [671]:
combined_taste.head()

Taste  Avg Rating  # of bars First/Second
0     distinct lemon         4.0          1  First Taste
1         ligt color         4.0          1  First Taste
2   simple red berry         4.0          1  First Taste
3              smoth         4.0          1  First Taste
4  cocoa and coconut         4.0          1  First Taste

In [672]:
#Filter out to keep only tastes that have 4 star rating
combined_best_taste=combined_taste[combined_taste['Avg Rating']>=4.0]

In [673]:
len(combined_best_taste)

27

_Combine worst tastes for first and second taste_

In [674]:
first_worst_taste=first_taste.tail(10)

In [675]:
second_worst_taste=second_taste.tail(10)

In [676]:
combined_worst_taste=first_worst_taste.append(second_worst_taste)

In [677]:
len(combined_worst_taste)

20

In [678]:
first_second_4_stars=first_second_taste[first_second_taste['Avg Rating']==4.0]
first_second_4_stars=first_second_4_stars.sort_values(by=['Avg Rating','# of bars'], ascending=[False, False]).reset_index()

_Plot these indiviual best tastes, individual worst tastes and best |combination of first, second taste_

In [680]:
# --- Plot best tastes ---

fig = px.bar(combined_best_taste, x="Taste", y="Avg Rating", color="First/Second",color_discrete_sequence=['rgb(82,188,163)','rgb(141,106,203)'])

fig=fig.update_layout(
    autosize=False,
    width=1000,
    height=500,
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    title_text="Best First and Second Tastes in chocolate Bars",
    xaxis_tickangle=-45,
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="darkblue",
    
    
)
fig=fig.update_traces(marker_line_color='#222A2A')
fig.show()

#---- Plot Worst tastes ---

fig1 = px.bar(combined_worst_taste, x="Taste", y="Avg Rating", color="First/Second",color_discrete_sequence=['rgb(176,226,209)',"rgb(203,213,232)"])

fig1=fig1.update_layout(
    autosize=False,
    width=900,
    height=500,
    
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    title_text="Worst First and Second Tastes in chocolate Bars",
    xaxis_tickangle=-45,
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="crimson",

)

fig1=fig1.update_traces(marker_line_color='#222A2A')
fig1.show()

#--- Table showing combination of First and Second taste of bars with rating ==4.0 --- 

fig2 = go.Figure(data=[go.Table(
    header=dict(values=list(["First Taste","Second Taste","Avg Rating", '# of Bars']),
                fill_color='paleturquoise',
                align='left'),
    cells=dict(values=[first_second_4_stars.first_taste, first_second_4_stars.second_taste, first_second_4_stars['Avg Rating'], first_second_4_stars['# of bars']],
               fill_color='lavender',
               align='left'))
])

fig2=fig2.update_layout(
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    title_text="First and Second Taste Combination for Chocolate with 4 star Ratings",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="darkmagenta",

)
fig2.show()

<div class="alert alert-block alert-info">

- We sorted out some top first tastes and top second tastes (based on ratings) irrespective of the other tastes that the chocolate bar might have.(First chart)
- One could mistakenly guess that if the first and second tastes of a bar belongs to one of the above shortlisted best tastes, then the taste or the rating of cholocate should be good enough, but from the combined analysis of first and second tastes in a chocolate, it looks like this is not True. 
- Even if one of the first and second tastes of a chocolate bar is marked as Top taste in either the first or second taste list, it is not necessary that the combination of those top tastes would result in a good chocolate. There is no fail-safe formular there. 
- It means that combination of <strong>tastes that do not show up in the top tastes</strong> could also result in a good cholocate bar.
- Above charts and the last tabular data supports this analysis of rating of combination of first and second taste VS rating of first or second tastes individually.

</div>

### Explore Ingredients distribution 


#### Cocoa Distribution

In [ ]:
cocoa_percent=chocolate.groupby('cocoa_percent')[['rating']].agg({'mean','count'})
cocoa_percent.columns=cocoa_percent.columns.droplevel(level=0)
cocoa_percent=cocoa_percent.rename(columns={'mean':'Avg Rating','count':'# of bars'})
cocoa_percent.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])
cocoa_percent
cocoa_percent=cocoa_percent.reset_index()

In [532]:
#Plot Cocoa distribution
fig3 = px.scatter(cocoa_percent, x="cocoa_percent", y="Avg Rating", 
                 size='# of bars',color_discrete_sequence=["#620042"])
fig3=fig3.update_layout(
    paper_bgcolor="snow",
    title_text="Cocoa % distribution and Ratings",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="brown",
    height=300
)
fig3.show()


<div class="alert alert-block alert-info">

- Evidently, most of the chocolates (904 chocoloates to be precise) consists 70% Cocoa and the average Rating is 3.2 stars which can be considered good)
- The trend of avg rating drops steadily after for Cocoa % > 70%. 
- From the graph it looks like people prefer or like chocolates whose Cocoa% is between 65% to 75% and most of the companies are aware of these statistics, hence number of chocolate bars produced within 65% to 75% Cocoa range is far more than that produced outside this range ! 

</div>

#### Other Ingredients Analysis

In [533]:
beans_percent=chocolate.groupby('beans')[['rating']].agg({'mean','count'})
beans_percent.columns=beans_percent.columns.droplevel(level=0)
beans_percent=beans_percent.rename(columns={'mean':'Avg Rating','count':'# of bars'})
beans_percent.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])
# No use of this colum - All chocolate bars have beans !

Avg Rating  # of bars
beans                           
have_bean    3.198561       2224

In [563]:
cocoa_butter_percent=chocolate.groupby('cocoa_butter')[['rating']].agg({'mean','count'})
cocoa_butter_percent.columns=cocoa_butter_percent.columns.droplevel(level=0)
cocoa_butter_percent=cocoa_butter_percent.rename(columns={'mean':'Avg Rating','count':'# of bars'})
cocoa_butter_percent=cocoa_butter_percent.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])

# Make this data ready for combining with other ingredients. 
cocoa_butter_percent=cocoa_butter_percent.reset_index()
cocoa_butter_percent.rename(columns={'cocoa_butter':'Ingredient'},inplace=True)
cocoa_butter_percent['Ingredient Present?']=['True','False']
cocoa_butter_percent['Ingredient']="Cocoa Butter"
cocoa_butter_percent


Ingredient  Avg Rating  # of bars Ingredient Present?
0  Cocoa Butter    3.202117       1535                True
1  Cocoa Butter    3.190639        689               False

In [551]:
vanilla_percent=chocolate.groupby('vanilla')[['rating']].agg({'mean','count'})
vanilla_percent.columns=vanilla_percent.columns.droplevel(level=0)
vanilla_percent=vanilla_percent.rename(columns={'mean':'Avg Rating','count':'# of bars'})
vanilla_percent=vanilla_percent.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])
vanilla_percent
# Make this data ready for combining with other ingredients. 

vanilla_percent=vanilla_percent.reset_index()
vanilla_percent.rename(columns={'vanilla':'Ingredient'},inplace=True)
vanilla_percent['Ingredient Present?']=['False','True']
vanilla_percent['Ingredient']="Vanilla"
vanilla_percent



Avg Rating  # of bars
vanilla                               
have_not_vanila    3.229398       1876
have_vanila        3.032328        348

Ingredient  Avg Rating  # of bars Ingredient Present?
0    Vanilla    3.229398       1876               False
1    Vanilla    3.032328        348                True

In [553]:
lecithin_percent=chocolate.groupby('lecithin')[['rating']].agg({'mean','count'})
lecithin_percent.columns=lecithin_percent.columns.droplevel(level=0)
lecithin_percent=lecithin_percent.rename(columns={'mean':'Avg Rating','count':'# of bars'})
lecithin_percent=lecithin_percent.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])
lecithin_percent

lecithin_percent=lecithin_percent.reset_index()
lecithin_percent.rename(columns={'lecithin':'Ingredient'},inplace=True)
lecithin_percent['Ingredient Present?']=['False','True']
lecithin_percent['Ingredient']="Lecithin"
lecithin_percent


Avg Rating  # of bars
lecithin                                
have_not_lecithin    3.214527       1745
have_lecithin        3.140397        479

Ingredient  Avg Rating  # of bars Ingredient Present?
0   Lecithin    3.214527       1745               False
1   Lecithin    3.140397        479                True

In [555]:
salt_percent=chocolate.groupby('salt')[['rating']].agg({'mean','count'})
salt_percent.columns=salt_percent.columns.droplevel(level=0)
salt_percent=salt_percent.rename(columns={'mean':'Avg Rating','count':'# of bars'})
salt_percent=salt_percent.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])
salt_percent=salt_percent.reset_index()
salt_percent

salt_percent.rename(columns={'salt':'Ingredient'},inplace=True)
salt_percent['Ingredient Present?']=['False','True']
salt_percent['Ingredient']="Salt"
salt_percent


salt  Avg Rating  # of bars
0  have_not_salt    3.201463       2187
1      have_salt    3.027027         37

Ingredient  Avg Rating  # of bars Ingredient Present?
0       Salt    3.201463       2187               False
1       Salt    3.027027         37                True

In [556]:
sugar_percent=chocolate.groupby('sugar')[['rating']].agg({'mean','count'})
sugar_percent.columns=sugar_percent.columns.droplevel(level=0)
sugar_percent=sugar_percent.rename(columns={'mean':'Avg Rating','count':'# of bars'})
sugar_percent=sugar_percent.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])
sugar_percent=sugar_percent.reset_index()
sugar_percent

sugar_percent.rename(columns={'sugar':'Ingredient'},inplace=True)
sugar_percent['Ingredient Present?']=['True','False']
sugar_percent['Ingredient']="Sugar"
sugar_percent


sugar  Avg Rating  # of bars
0      have_sugar    3.206396       2142
1  have_not_sugar    2.993902         82

Ingredient  Avg Rating  # of bars Ingredient Present?
0      Sugar    3.206396       2142                True
1      Sugar    2.993902         82               False

In [74]:
sugar_percent2=chocolate.groupby('sweetener_without_sugar')[['rating']].agg({'mean','count'})
sugar_percent2.columns=sugar_percent2.columns.droplevel(level=0)
sugar_percent2=sugar_percent2.rename(columns={'mean':'Avg Rating','count':'# of bars'})
sugar_percent2.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])
# This one gives similar result as sugar. So not including this one in analysis.

Avg Rating  # of bars
sweetener_without_sugar                                
have_not_sweetener_without_sugar    3.205703       2148
have_sweetener_without_sugar        2.996711         76

In [564]:
#combine these all above and then plot a graph and draw conclusions
combined_ingredients=cocoa_butter_percent.append(salt_percent)
combined_ingredients=combined_ingredients.append(sugar_percent)
combined_ingredients=combined_ingredients.append(lecithin_percent)
combined_ingredients=combined_ingredients.append(vanilla_percent)

In [565]:
combined_ingredients

Ingredient  Avg Rating  # of bars Ingredient Present?
0  Cocoa Butter    3.202117       1535                True
1  Cocoa Butter    3.190639        689               False
0          Salt    3.201463       2187               False
1          Salt    3.027027         37                True
0         Sugar    3.206396       2142                True
1         Sugar    2.993902         82               False
0      Lecithin    3.214527       1745               False
1      Lecithin    3.140397        479                True
0       Vanilla    3.229398       1876               False
1       Vanilla    3.032328        348                True

In [589]:
# Plot the presence/absence of these ingredients vs Avg. Rating
fig4 = go.Figure()
fig4=fig4.add_trace(go.Bar(
    x=combined_ingredients[combined_ingredients['Ingredient Present?']=='True']['Ingredient'],
    y=combined_ingredients[combined_ingredients['Ingredient Present?']=='True']['Avg Rating'],
    hovertext=combined_ingredients[combined_ingredients['Ingredient Present?']=='True']['# of bars'],
    hovertemplate = 'Ingredient : %{x}' +  '<br>Avg Rating: %{y}'+'<br>Number of Chocolate bars : %{hovertext}' ,
    name='Ingredient Present',
    marker_color='rgb(56, 166, 165)'
))
fig4=fig4.add_trace(go.Bar(
    x=combined_ingredients[combined_ingredients['Ingredient Present?']=='False']['Ingredient'],
    y=combined_ingredients[combined_ingredients['Ingredient Present?']=='False']['Avg Rating'],
    hovertext=combined_ingredients[combined_ingredients['Ingredient Present?']=='False']['# of bars'],
    hovertemplate = 'Ingredient : %{x}' +  '<br>Avg Rating: %{y}'+'<br>Number of Chocolate bars : %{hovertext}' ,
    name='Ingredient Absent',
    marker_color='rgb(55, 83, 109)'
))

fig4=fig4.update_layout(barmode='group',
                       paper_bgcolor="snow",
                    plot_bgcolor="snow",
    title_text="Ingredients vs Rating",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="black",
    bargap=0.25, # gap between bars of adjacent location coordinates.
    bargroupgap=0.05, # gap between bars of the same location coordinate.
    height=500,
    width=700,
                       
    yaxis=dict(
        title='Avg rating in stars',
        titlefont_size=12,
        tickfont_size=12,
    ),                   )
fig4.show()






<div class="alert alert-block alert-info">

- From above data, it seems that Cocoa Butter and Sugar are preferred whereas Salt, Lecithin and Vanilla are not preferred in chocolate bars. 
- The number of chocolate bars and Avg CSAT with Salt, Lecithin and Vanilla present is less than when when these ingredients are not present.

</div>

### Countries producing chocolate BEAN, their ratings and the number of beans that they produce


In [616]:
#Load country code for map
#Downloaded country code from : https://www.kaggle.com/juanumusic/countries-iso-codes
country_code=pd.read_csv("country_code.csv")

In [681]:
bean_origin=chocolate.groupby('country_of_bean_origin_new')[['rating']].agg({'mean','count'})
bean_origin.columns=bean_origin.columns.droplevel(level=0)
bean_origin=bean_origin.rename(columns={'mean':'Avg Rating','count':'# of bars'})
bean_origin=bean_origin.sort_values(by=['Avg Rating','# of bars'],ascending=[False,False])
bean_origin=bean_origin.reset_index()
bean_origin

country_of_bean_origin_new  Avg Rating  # of bars
0                    Solomon Islands    3.450000         10
1                              Congo    3.318182         11
2                           Thailand    3.300000          5
3                               Cuba    3.291667         12
4                          Guatemala    3.273585         53
5                   Papua New Guinea    3.270833         48
6                              Haiti    3.270833         24
7                         Madagascar    3.264331        157
8                            Vietnam    3.253906         64
9                             Brazil    3.250000         74
10                         Australia    3.250000          3
11                             Gabon    3.250000          1
12                          Suriname    3.250000          1
13                         Nicaragua    3.239130         92
14                            Belize    3.238462         65
15                         Venezuela    3.233193        238
16                          Tanzania    3.230159         63
17                          Honduras    3.228261         23
18               Trinidad and Tobago    3.217949         39
19                          Colombia    3.215385         65
20                           Ecuador    3.210052        194
21                Dominican Republic    3.205500        200
22                              Peru    3.198068        207
23                            U.S.A.    3.187500         28
24                           Bolivia    3.180556         72
25                           Jamaica    3.173913         23
26                         Indonesia    3.157895         19
27                             Ghana    3.156250         32
28                            Panama    3.156250          8
29                        Costa Rica    3.154762         42
30                             India    3.137500         20
31                            Mexico    3.111111         45
32                       Philippines    3.105263         19
33                       Saint Lucia    3.083333          9
34                          Cameroon    3.083333          3
35                           Liberia    3.083333          3
36                             Samoa    3.083333          3
37                              Togo    3.083333          3
38                             Blend    3.082143        140
39             Sao Tome and Principe    3.078125         16
40                              Fiji    3.066667         15
41                           Vanuatu    3.062500         12
42                          Malaysia    3.035714          7
43                           Grenada    3.014706         17
44                       El Salvador    3.000000          6
45      Democratic Republic of Congo    3.000000          1
46                           Myanmar    3.000000          1
47                            Uganda    2.875000         10
48                           Nigeria    2.875000          2
49                         Sri Lanka    2.875000          2
50                            Taiwan    2.875000          2
51                      Sierra Leone    2.833333          3
52                       Puerto Rico    2.750000          6
53                       Ivory Coast    2.750000          4
54                        Martinique    2.750000          1
55  Saint Vincent and the Grenadines    2.750000          1

In [686]:
#Join bean_origin dataframe with the country code
bean_origin=bean_origin.merge(country_code[['English short name lower case','Alpha-3 code']], left_on='country_of_bean_origin_new',right_on='English short name lower case', how='left')

In [687]:
import plotly.graph_objects as go
import pandas as pd

#--- Following code plots avg rating of the beans produced in the given countries ---

fig = go.Figure(data=go.Choropleth(
    locations = bean_origin['Alpha-3 code'],
    z = bean_origin['Avg Rating'],
    text = bean_origin['country_of_bean_origin_new'],
    colorscale = 'darkmint',
    autocolorscale=False,
    reversescale=False,
    marker_line_color='aqua',
    marker_line_width=0.5,
    colorbar_ticksuffix = ' stars',
    colorbar_title = 'Avg Rating',
))

fig=fig.add_trace(go.Scattergeo(
    lon = [-89,90],
    lat = [-0.01,0.01],
    hoverinfo = 'text',
    text = 'Equator',
    mode = 'lines',
    showlegend = False,
    marker = dict(
        size = 2,
        color = 'goldenrod',
        
    )))
fig=fig.add_trace(go.Scattergeo(
    lon = [-90,91],
    lat = [-0.01,0.01],
    hoverinfo = 'text',
    text = 'Equator',
    mode = 'lines',
    showlegend = False,
    marker = dict(
        size = 2,
        color = 'goldenrod',
        
        
    )))
fig.update_layout(
    title_text='Country Of Bean Orgin And Avg Rating Of The Cholocate Bar',
    height=500,
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="darkolivegreen",
    geo=dict(
        showframe=True,
        showcoastlines=False,
        projection_type='natural earth'
    )
)
 
    # Above code plots first graph.
    #--------------------  
 ## Now plot country of bean origin and number of chocolate beans produced to check which country produces the most bean  
 # Below code plots second graph
    
fig1 = go.Figure(data=go.Choropleth(
    locations = bean_origin['Alpha-3 code'],
    z = bean_origin['# of bars'],
    text = bean_origin['country_of_bean_origin_new'],
    colorscale = 'ice',
    autocolorscale=False,
    reversescale=True,
    marker_line_color='mediumturquoise',
    marker_line_width=0.5,
    colorbar_ticksuffix = ' types',
    colorbar_title = '# Beans Produced',
))

fig1=fig1.add_trace(go.Scattergeo(
    lon = [-89,90],
    lat = [-0.01,0.01],
    hoverinfo = 'text',
    text = 'Equator',
    mode = 'lines',
    showlegend = False,
    marker = dict(
        size = 2,
        color = 'goldenrod',
        
    )))
fig1=fig1.add_trace(go.Scattergeo(
    lon = [-90,91],
    lat = [-0.01,0.01],
    hoverinfo = 'text',
    text = 'Equator',
    mode = 'lines',
    showlegend = False,
    marker = dict(
        size = 2,
        color = 'goldenrod',
        
        
    )))
fig1.update_layout(
    title_text='Country Of Bean Orgin And Number of Chocolate Beans Produced',
    height=500,
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="darkolivegreen",
    geo=dict(
        showframe=True,
        showcoastlines=False,
        projection_type='natural earth'
    )
)
#https://plotly.com/python/choropleth-maps/ ---- Reference 


<div class="alert alert-block alert-info">

- Australia, some small countries like Soloman Islands,Thailand, Ghana, Congo, Madagascar, Vietnam and most of the countries in South America produce good quality chocolate beans. (Avg rating of 3.25 and above) 
- From the first graph, it looks like all the beans are produced in countries near the Equator. As you go north or south from the equator, you would hardly find any countries producing chocolating beans in north or south. (Maybe climatic conditions near equator are favorable for chocolate bean cultivation)
- This is also supported by the data depicted in second graph : Countries like Peru, Venezuela, Brazil, Bolivia, Columbia, Nicaragua,Tanzania, Ecuador most of the chocolate beans. These are countries that are either very close to equator or has equator passing through them. 
</div>


### Chocolate BAR Production Analysis 


In [688]:
company_loc=chocolate.groupby('company_location_new')[['rating']].agg({'mean','count'})
company_loc.columns=company_loc.columns.droplevel(level=0)
company_loc=company_loc.rename(columns={'mean':'Avg Rating','count':'# of bars'}).reset_index()
company_loc.sort_values(by=['# of bars','Avg Rating'],ascending=[False,False])


company_location_new  Avg Rating  # of bars
58                             U.S.A    3.175897        948
17                            France    3.263473        167
6                             Canada    3.306962        158
57                              U.K.    3.111742        132
27                             Italy    3.250000         67
3                            Belgium    3.187500         56
1                          Australia    3.367647         51
13                           Ecuador    3.060000         50
53                       Switzerland    3.313953         43
18                           Germany    3.256410         39
50                             Spain    3.264286         35
11                           Denmark    3.338710         31
60                         Venezuela    3.112903         31
2                            Austria    3.258333         30
35                       New Zealand    3.240385         26
28                             Japan    3.201923         26
23                           Hungary    3.200000         25
8                           Colombia    3.187500         24
5                             Brazil    3.271739         23
34                       Netherlands    3.237500         20
38                              Peru    3.065789         19
61                           Vietnam    3.333333         15
47                         Singapore    3.153846         13
30                        Madagascar    3.134615         13
12                Dominican Republic    3.122727         11
21                         Guatemala    3.350000         10
36                         Nicaragua    3.100000         10
54                            Taiwan    3.100000         10
26                            Israel    3.250000          9
9                         Costa Rica    3.138889          9
49                       South Korea    3.156250          8
29                         Lithuania    3.125000          8
33                            Mexico    2.968750          8
40                            Poland    3.500000          7
46             Sao Tome and Principe    3.107143          7
22                          Honduras    3.208333          6
52                            Sweden    3.000000          6
55                          Thailand    3.300000          5
39                       Philippines    3.150000          5
0                          Argentina    3.500000          4
56                            U.A.E.    3.375000          4
24                           Iceland    3.312500          4
15                              Fiji    3.250000          4
48                      South Africa    2.750000          4
59                           Vanuatu    2.750000          4
42                       Puerto Rico    2.625000          4
37                            Norway    3.333333          3
10                    Czech Republic    3.000000          3
14                       El Salvador    3.000000          3
20                           Grenada    2.833333          3
31                          Malaysia    2.833333          3
41                          Portugal    2.750000          3
44                       Saint Lucia    2.750000          3
7                              Chile    3.750000          2
4                            Bolivia    3.250000          2
16                           Finland    3.250000          2
25                             India    2.625000          2
43                            Russia    3.250000          1
51                          Suriname    3.250000          1
19                             Ghana    2.750000          1
32                        Martinique    2.750000          1
45  Saint Vincent and the Grenadines    2.750000          1
62                             Wales    2.750000          1

In [689]:
company_loc=company_loc.merge(country_code[['English short name lower case','Alpha-3 code']], left_on='company_location_new',right_on='English short name lower case', how='left')

In [690]:
top_20_company_loc=company_loc.sort_values(by='# of bars',ascending=False).head(20)

In [691]:
top_20_company_loc # Took top 20 companies to be able to plot it properly since the data is unevenly distributed.

company_location_new  Avg Rating  # of bars English short name lower case  \
58                U.S.A    3.175897        948                         U.S.A   
17               France    3.263473        167                        France   
6                Canada    3.306962        158                        Canada   
57                 U.K.    3.111742        132                          U.K.   
27                Italy    3.250000         67                         Italy   
3               Belgium    3.187500         56                       Belgium   
1             Australia    3.367647         51                     Australia   
13              Ecuador    3.060000         50                       Ecuador   
53          Switzerland    3.313953         43                   Switzerland   
18              Germany    3.256410         39                       Germany   
50                Spain    3.264286         35                         Spain   
11              Denmark    3.338710         31                       Denmark   
60            Venezuela    3.112903         31                     Venezuela   
2               Austria    3.258333         30                       Austria   
28                Japan    3.201923         26                         Japan   
35          New Zealand    3.240385         26                   New Zealand   
23              Hungary    3.200000         25                       Hungary   
8              Colombia    3.187500         24                      Colombia   
5                Brazil    3.271739         23                        Brazil   
34          Netherlands    3.237500         20                   Netherlands   

   Alpha-3 code  
58          USA  
17          FRA  
6           CAN  
57          GBR  
27          ITA  
3           BEL  
1           AUS  
13          ECU  
53          CHE  
18          DEU  
50          ESP  
11          DNK  
60          VEN  
2           AUT  
28          JPN  
35          NZL  
23          HUN  
8           COL  
5           BRA  
34          NLD

In [692]:
import plotly.graph_objects as go
import pandas as pd


fig = go.Figure(data=go.Choropleth(
    locations = company_loc['Alpha-3 code'],
    z = company_loc['Avg Rating'],
    text = company_loc['company_location_new'],
    colorscale = 'darkmint',
    autocolorscale=False,
    reversescale=False,
    marker_line_color='aqua',
    marker_line_width=0.5,
    colorbar_ticksuffix = ' stars',
    colorbar_title = 'Avg Rating',
))

fig=fig.add_trace(go.Scattergeo(
    lon = [-89,90],
    lat = [-0.01,0.01],
    hoverinfo = 'text',
    text = 'Equator',
    mode = 'lines',
    showlegend = False,
    marker = dict(
        size = 2,
        color = 'goldenrod',
        
    )))
fig=fig.add_trace(go.Scattergeo(
    lon = [-90,91],
    lat = [-0.01,0.01],
    hoverinfo = 'text',
    text = 'Equator',
    mode = 'lines',
    showlegend = False,
    marker = dict(
        size = 2,
        color = 'goldenrod',
        
        
    )))
fig.update_layout(
    title_text='Country of Company Location And Avg Rating Of The Cholocate Bar',
    height=500,
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="darkolivegreen",
    geo=dict(
        showframe=True,
        showcoastlines=False,
        projection_type='natural earth',
        
    )
)
 
    # Abive code plots first graph.
    #--------------------  
 ## Now plot country of bean origin and number of chocolate beans produced to check which country produces the most chocolates 
 # Below code plots second graph
    
fig1 = go.Figure(data=go.Choropleth(
    locations = top_20_company_loc['Alpha-3 code'],
    z = top_20_company_loc['# of bars'],
    text = top_20_company_loc['company_location_new'],
    colorscale = 'deep',
    #autocolorscale=True,
    reversescale=False,
    marker_line_color='mediumturquoise',
    marker_line_width=0.5,
    colorbar_ticksuffix = ' types',
    colorbar_title = '# Chocolate Bars Produced',
))

fig1=fig1.add_trace(go.Scattergeo(
    lon = [-89,90],
    lat = [-0.01,0.01],
    hoverinfo = 'text',
    text = 'Equator',
    mode = 'lines',
    showlegend = False,
    marker = dict(
        size = 2,
        color = 'goldenrod',
        
    )))
fig1=fig1.add_trace(go.Scattergeo(
    lon = [-90,91],
    lat = [-0.01,0.01],
    hoverinfo = 'text',
    text = 'Equator',
    mode = 'lines',
    showlegend = False,
    marker = dict(
        size = 2,
        color = 'goldenrod',
        
        
    )))
fig1.update_layout(
    title_text='Country of Company Location And Number of Chocolate Bars Produced',
    height=500,
    paper_bgcolor="snow",
    plot_bgcolor="snow",
    font_family="Arial",
    title_font_family="Times New Roman",
    title_font_color="darkolivegreen",
    geo=dict(
        showframe=True,
        showcoastlines=False,
        projection_type='natural earth'
    )
)
#https://plotly.com/python/choropleth-maps/ ---- Reference 


<div class="alert alert-block alert-info">

- After learning aboout the countries that lead in chocolate bean production, one would think that those countries would also lead in producing the chocolate bars, however, on the countrary, these countries produce less variety of chocolate bars in comparison to Countries like USA, UK, France, Italy, Belgium, Australia.
- Infact, it seems like countries like Russia, Canada, UK where there is no chocolate bean production, manages to import these beans and produce good quality chocolate bar.
- The second chart shows the top 20 countries that produces most number of chocolate bars.
- U.S.A clearly produces a LOT of chocolate bars (around 948) and is followed by Canada, France, UK, Italy. 
- Some of the countries like Madagascar, Bolivia that produces a lot and good quality of chocoate beans ranks below 20 in the number of chocolate bar producer countries, thus justifying the fact that cultivation of chocolate bean is not related to the production of chocoate bars in that region 
</div>
